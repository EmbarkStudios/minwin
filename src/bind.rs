mod emit;
mod gather;

pub use gather::{Disambiguate, Impls, Item};

use anyhow::Context as _;
use rayon::prelude::*;
use windows_metadata::reader as wmr;

pub type InterfaceMap = std::collections::BTreeMap<String, BTreeSet<String>>;

/// All of the unique items gathered based on a root list of names
#[derive(Clone)]
pub struct ItemSet {
    /// Set of structs, unions, interfaces, classes, enums, type aliases, and function pointers
    pub types: BTreeMap<wmr::Type, Impls>,
    /// Set of functions
    pub functions: BTreeSet<wmr::MethodDef>,
    /// Set of constants, including individual enum values
    pub constants: BTreeSet<wmr::Field>,
}

fn qualify_item(reader: &wmr::Reader, bname: String) -> anyhow::Result<(String, Item)> {
    let (mut item, bname) = Item::parse(bname)?;

    if bname.starts_with("Windows.") {
        Ok((bname, item))
    } else {
        for ns in reader.namespaces() {
            // We could cheat and take casing into account, but there
            // are of course exceptions, and it's fast enough to just
            // brute force it

            // We _do_ consider which items are most likely to be bound,
            // namely functions -> constants/enums -> records
            let mut find = || {
                if matches!(item.dis, Disambiguate::Any | Disambiguate::Function)
                    && reader
                        .namespace_functions(ns)
                        .any(|f| reader.method_def_name(f) == bname)
                {
                    item.dis = Disambiguate::Function;
                    return true;
                }

                if matches!(item.dis, Disambiguate::Any | Disambiguate::Constant)
                    && reader
                        .namespace_constants(ns)
                        .any(|c| reader.field_name(c) == bname)
                {
                    item.dis = Disambiguate::Constant;
                    return true;
                }

                reader.namespace_types(ns, &Default::default()).any(|td| {
                    if reader.type_def_kind(td) == wmr::TypeKind::Enum {
                        if reader
                            .type_def_fields(td)
                            .any(|f| reader.field_name(f) == bname)
                        {
                            item.dis = Disambiguate::Constant;
                            return true;
                        }
                    } else if reader.type_def_name(td) == bname {
                        return true;
                    }

                    false
                })
            };

            if find() {
                let qualified = format!("{ns}.{bname}");
                tracing::debug!("resolved {bname} => {qualified}");
                return Ok((qualified, item));
            }
        }

        anyhow::bail!("unable to resolve '{bname}'");
    }
}

pub fn qualify_items(
    reader: &wmr::Reader,
    items: Vec<String>,
) -> anyhow::Result<Vec<(String, Item)>> {
    // Allow the user to specify simple or fully qualified names
    items
        .into_par_iter()
        .map(|bname| qualify_item(reader, bname))
        .collect()
}

use std::{
    collections::{BTreeMap, BTreeSet},
    fmt,
};
pub struct Bucket<T> {
    pub items: BTreeSet<T>,
    pub num: u32,
}

impl<T: std::cmp::Ord> Bucket<T> {
    #[inline]
    fn new(items: BTreeSet<T>) -> Self {
        Self { items, num: 0 }
    }

    #[inline]
    fn has(&self, item: &T) -> bool {
        self.items.contains(item)
    }

    #[inline]
    fn inc(&mut self) {
        self.num += 1;
    }
}

pub struct Items {
    /// Type aliases, not super expensive, so worth differentiating from other types
    pub aliases: Bucket<wmr::Type>,
    /// Functions
    pub functions: Bucket<wmr::MethodDef>,
    /// Constants, both global and enum variants
    pub constants: Bucket<wmr::Field>,
    /// Structs and unions
    pub records: Bucket<wmr::TypeDef>,
    /// Function pointers, rare
    pub func_pointers: Bucket<wmr::TypeDef>,
    /// COM interfaces, only present when emitting COM
    pub interfaces: Bucket<wmr::TypeDef>,
}

/// Determines the style used to emit extern functions from Win32
#[derive(Copy, Clone, PartialEq, Eq, Default, serde::Deserialize)]
pub enum LinkingStyle {
    /// Emits normal `#[link(name = "<dll>")]` extern blocks collated by dynamic
    /// library.
    ///
    /// There are exactly 5 non-dll dynamic libraries in the Win32 bindings, this
    /// style won't work with them, and you will get a panic about it
    Normal,
    /// Emits 1 `windows_targets::link!` per extern function
    #[default]
    WindowsTargets,
    /// Emits extern blocks similarly to [`Self::Normal`], except each one is
    /// `kind = "raw-dylib"`.
    ///
    /// This allows linking without needing import libs, but requires 1.65 for
    /// `x86_64` and `aarch64`, and is currently unstable on `x86`, though will
    /// be stable in (presumably) 1.71 <https://github.com/rust-lang/rust/pull/109677>
    RawDylib,
}

impl fmt::Display for LinkingStyle {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Normal => f.write_str("normal"),
            Self::WindowsTargets => f.write_str("targets"),
            Self::RawDylib => f.write_str("rawdylib"),
        }
    }
}

/// Determines the style used to emit enums
#[derive(Copy, Clone, PartialEq, Eq, Default, serde::Deserialize)]
pub enum EnumStyle {
    /// Emits bindgen style enums
    ///
    /// ```
    /// pub type ENUM_NAME = u32;
    /// pub const ENUM_VALUE_ONE: ENUM_NAME = 1;
    /// pub const ENUM_VALUE_TWO: ENUM_NAME = 2;
    ///
    /// fn function_that_takes_enum(enum_val: ENUM_NAME) {}
    ///
    /// fn main() {
    ///     function_that_takes_enum(ENUM_VALUE_ONE);
    /// }
    /// ```
    #[default]
    Bindgen,
    /// Emits minwin style enums
    ///
    /// ```
    /// pub mod ENUM_NAME {
    ///     pub type Enum = u32;
    ///     pub const ENUM_VALUE_ONE: Enum = 1;
    ///     pub const ENUM_VALUE_TWO: Enum = 2;
    /// }
    ///
    /// fn function_that_takes_enum(enum_val: ENUM_NAME::Enum) {}
    ///
    /// fn main() {
    ///     function_that_takes_enum(ENUM_NAME::ENUM_VALUE_ONE);
    /// }
    /// ```
    Minwin,
}

impl fmt::Display for EnumStyle {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Bindgen => f.write_str("bindgen"),
            Self::Minwin => f.write_str("minwin"),
        }
    }
}

#[derive(serde::Deserialize, Copy, Clone)]
#[serde(rename_all = "kebab-case")]
pub struct MinwinBindConfig {
    /// The linking style to use for extern functions
    pub linking_style: LinkingStyle,
    /// The style of enums emitted, affects both declaration and usage
    pub enum_style: EnumStyle,
    /// If true, the `windows-core` crate is used for various core types
    /// such as `HRESULT`
    pub use_core: bool,
    /// If true, identifiers are fixed to remove pointless Hungarian notation
    pub fix_naming: bool,
    /// If true, the casing of all items will be changed to follow Rust casing conventions
    pub use_rust_casing: bool,
    /// If true, formats the output
    pub pretty_print: bool,
    /// If true, emits a version header at the beginning of the bindings
    pub add_version_header: bool,
}

impl Default for MinwinBindConfig {
    fn default() -> Self {
        Self {
            linking_style: Default::default(),
            enum_style: Default::default(),
            use_core: false,
            fix_naming: false,
            use_rust_casing: false,
            pretty_print: true,
            // We don't want this in tests
            add_version_header: !cfg!(debug_assertions),
        }
    }
}

#[derive(serde::Deserialize)]
pub enum BindConfig {
    /// Uses `windows-bindgen` to generate bindings.
    ///
    /// This option gives limited configuration and no way to inspect the number
    /// and types of items emitted. Requires `rustfmt` is installed.
    ///
    /// Use this if you want the "official" bindings.
    Bindgen,
    /// Uses `minwin` to generate bindings.
    ///
    /// This option allows for various customizations to be applied to the
    /// emitted bindings, as well as complete introspection on all of the emitted
    /// items
    Minwin(MinwinBindConfig),
}

impl fmt::Display for BindConfig {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Bindgen => f.write_str("bindgen"),
            Self::Minwin(mwbc) => {
                f.write_fmt(format_args!(
                    "ls-{}_es-{}",
                    mwbc.linking_style, mwbc.enum_style
                ))?;

                if mwbc.use_core {
                    f.write_str("_core")?;
                }

                if mwbc.fix_naming {
                    f.write_str("_fixnaming")?;
                }

                if mwbc.use_rust_casing {
                    f.write_str("_rustcasing")?;
                }

                Ok(())
            }
        }
    }
}

pub struct BindOutput {
    /// The full set of generated bindings
    pub bindings: String,
    /// The full set of items that were emitted. Not available if [`BindConfig::Bindgen`] is used.
    pub items: Option<BTreeMap<String, Items>>,
}

pub fn bind<S: Into<String>>(
    names: impl IntoIterator<Item = S>,
    interfaces: BTreeMap<String, BTreeSet<String>>,
    config: BindConfig,
) -> anyhow::Result<BindOutput> {
    let names: Vec<_> = names.into_iter().map(|n| n.into()).collect();
    anyhow::ensure!(
        !names.is_empty() || !interfaces.is_empty(),
        "1 or more items must be specified to bind"
    );

    let files = &wmr::File::with_default(&[]).unwrap();
    let reader = &wmr::Reader::new(files);

    let items = qualify_items(reader, names)?;
    let ifaces = interfaces
        .into_par_iter()
        .map(|(name, methods)| qualify_item(reader, name).map(|(name, _)| (name, methods)))
        .collect::<anyhow::Result<BTreeMap<_, _>>>()?;

    anyhow::ensure!(
        !items.is_empty() || !ifaces.is_empty(),
        "unable to locate any items to bind"
    );

    match config {
        BindConfig::Bindgen => {
            let items: Vec<_> = items
                .iter()
                .map(|(n, _i)| n.as_str())
                .chain(ifaces.keys().map(|k| k.as_str()))
                .collect();
            Ok(BindOutput {
                bindings: windows_bindgen::standalone(&items),
                items: None,
            })
        }
        BindConfig::Minwin(config) => {
            let gatherer =
                gather::Gatherer::new(reader, items.iter().map(|(n, i)| (n.as_str(), *i)), &ifaces);
            let items = gatherer.gather();

            let emit = emit::Emit {
                items,
                reader,
                ifaces,
                layouts: Some(emit::load_clang_layouts()),
                config,
            };

            do_minwin(emit)
        }
    }
}

pub fn do_minwin(emit: emit::Emit<'_>) -> anyhow::Result<BindOutput> {
    // Gather stats now that we have all of the items we want to emit
    let mut ns_items = BTreeMap::<String, Items>::new();
    let reader = emit.reader;

    let index_namespace = |name: &str, ns_items: &mut BTreeMap<String, Items>| {
        assert!(
            reader.namespaces().any(|ns| ns == name),
            "namespace {name} was not found"
        );

        let functions = reader.namespace_functions(name).collect();
        let mut constants: BTreeSet<_> = reader.namespace_constants(name).collect();

        let mut aliases = BTreeSet::new();
        let mut records = BTreeSet::new();
        let mut interfaces = BTreeSet::new();
        let mut func_pointers = BTreeSet::new();

        for ty in reader.namespace_types(name, &Default::default()) {
            let kind = reader.type_def_kind(ty);

            match kind {
                wmr::TypeKind::Struct => {
                    if reader.type_def_is_handle(ty) {
                        aliases.insert(wmr::Type::TypeDef((ty, Vec::new())));
                    } else {
                        records.insert(ty);
                    }
                }
                wmr::TypeKind::Enum => {
                    aliases.insert(wmr::Type::TypeDef((ty, Vec::new())));
                    constants.extend(reader.type_def_fields(ty));
                }
                wmr::TypeKind::Delegate => {
                    func_pointers.insert(ty);
                }
                wmr::TypeKind::Class | wmr::TypeKind::Interface => {
                    let ty_name = reader.type_def_name(ty);
                    let qname = format!("{name}.{ty_name}");
                    if emit.ifaces.contains_key(&qname) {
                        interfaces.insert(ty);
                    } else {
                        aliases.insert(wmr::Type::TypeDef((ty, Vec::new())));
                    }
                }
            }
        }

        let items = Items {
            functions: Bucket::new(functions),
            constants: Bucket::new(constants),
            records: Bucket::new(records),
            aliases: Bucket::new(aliases),
            func_pointers: Bucket::new(func_pointers),
            interfaces: Bucket::new(interfaces),
        };

        ns_items.insert(name.to_owned(), items);
    };

    for func in emit.items.functions.iter().cloned() {
        let items = match ns_items
            .iter_mut()
            .find(|(_ns, items)| items.functions.has(&func))
        {
            Some(nsi) => nsi.1,
            None => {
                let ns = reader
                    .namespaces()
                    .find(|ns| reader.namespace_functions(ns).any(|m| m == func))
                    .with_context(|| {
                        format!(
                            "unable to find namespace containing function {}",
                            reader.method_def_name(func)
                        )
                    })?;

                index_namespace(ns, &mut ns_items);
                ns_items.get_mut(ns).unwrap()
            }
        };

        items.functions.inc();
    }

    for constant in emit.items.constants.iter().cloned() {
        let items = match ns_items
            .iter_mut()
            .find(|(_ns, items)| items.constants.has(&constant))
        {
            Some(nsi) => nsi.1,
            None => {
                let ns = if let wmr::Type::TypeDef((td, _)) = reader.field_type(constant, None) {
                    reader.type_def_namespace(td)
                } else {
                    reader
                        .namespaces()
                        .find(|ns| reader.namespace_constants(ns).any(|c| c == constant))
                        .with_context(|| {
                            format!(
                                "unable to find namespace containing constant {}",
                                reader.field_name(constant)
                            )
                        })?
                };

                index_namespace(ns, &mut ns_items);
                ns_items.get_mut(ns).unwrap()
            }
        };

        items.constants.inc();
    }

    for ty in emit.items.types.keys() {
        // Just ignore the core types, they're not a big deal
        let wmr::Type::TypeDef((td, _)) = &ty else { continue; };
        let td = *td;

        let ns = reader.type_def_namespace(td);
        if !ns_items.contains_key(ns) {
            index_namespace(ns, &mut ns_items);
        }

        let items = ns_items.get_mut(ns).with_context(|| {
            format!(
                "failed to index namespace '{ns}' to locate typedef '{}'",
                reader.type_def_name(td)
            )
        })?;

        match reader.type_def_kind(td) {
            wmr::TypeKind::Enum => items.aliases.inc(),
            wmr::TypeKind::Struct => items.records.inc(),
            wmr::TypeKind::Class | wmr::TypeKind::Interface => {
                let ty_name = reader.type_def_name(td);
                let qname = format!("{ns}.{ty_name}");
                if emit.ifaces.contains_key(&qname) {
                    items.interfaces.inc();
                } else {
                    items.aliases.inc();
                }
            }
            wmr::TypeKind::Delegate => items.func_pointers.inc(),
        }
    }

    let bindings = emit.emit()?;

    Ok(BindOutput {
        bindings,
        items: Some(ns_items),
    })
}
