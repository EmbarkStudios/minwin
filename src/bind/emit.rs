mod constant;
mod function;
mod function_pointer;
mod ostream;
mod record;
mod shared;

use ostream::OutputStream;
use shared::*;

use super::{Impls, InterfaceMap};
use anyhow::Context as _;
use proc_macro2::{self as pm, TokenStream};
use quote::quote;
use std::collections::BTreeMap;
use windows_metadata::reader::{self as wmr, Field, Reader, Type, TypeDef, TypeKind};

pub struct Emit<'r> {
    /// The actual items we are emitting
    pub items: super::ItemSet,
    /// Used to query the metadata database
    pub reader: &'r Reader<'r>,
    /// List of interfaces to emit, if an interface/class is not in this list it
    /// is emitted as a void* type alias instead
    pub ifaces: InterfaceMap,
    /// We collect additional layout information not currently present in the
    /// the metadata to supplement it
    pub layouts: Option<&'static BTreeMap<String, ArchLayouts>>,
    pub config: super::MinwinBindConfig,
}

impl<'r> Emit<'r> {
    pub fn emit(self) -> anyhow::Result<String> {
        let mut os = OutputStream::new(self.reader);

        // Insert simple header to let reviewers etc know where these bindings came from
        if self.config.add_version_header {
            let vs_header = format!(
                " Bindings generated by `minwin` {}",
                std::env!("CARGO_PKG_VERSION")
            );

            os.root.extend(quote! { #![doc = #vs_header] });
        }

        os.root.extend(quote! {
            #![allow(non_snake_case, non_upper_case_globals, non_camel_case_types, clippy::upper_case_acronyms)]
        });

        let reader = self.reader;

        for (ty, impls) in &self.items.types {
            if !self.config.use_core && !matches!(ty, Type::TypeDef(..)) {
                let (ident, ts) = match ty {
                    Type::HRESULT => {
                        let ident = self.config.make_ident("HRESULT", IdentKind::Type);
                        let ts = quote! { pub type #ident = i32; };
                        (ident, ts)
                    }
                    Type::String => {
                        let ident = self.config.make_ident("HSTRING", IdentKind::Type);
                        let ts = quote! { pub type #ident = *mut ::core::ffi::c_void; };
                        (ident, ts)
                    }
                    Type::IUnknown => {
                        let ident = self.config.make_ident("IUnknown", IdentKind::Type);
                        let ts = quote! { pub type #ident = *mut ::core::ffi::c_void; };
                        (ident, ts)
                    }
                    Type::IInspectable => {
                        let ident = self.config.make_ident("IInspectable", IdentKind::Type);
                        let ts = quote! { pub type #ident = *mut ::core::ffi::c_void; };
                        (ident, ts)
                    }
                    Type::PSTR => {
                        let ident = self.config.make_ident("PSTR", IdentKind::Type);
                        let ts = quote! { pub type #ident = *mut u8; };
                        (ident, ts)
                    }
                    Type::PWSTR => {
                        let ident = self.config.make_ident("PWSTR", IdentKind::Type);
                        let ts = quote! { pub type #ident = *mut u16; };
                        (ident, ts)
                    }
                    Type::PCSTR => {
                        let ident = self.config.make_ident("PCSTR", IdentKind::Type);
                        let ts = quote! { pub type #ident = *const u8; };
                        (ident, ts)
                    }
                    Type::PCWSTR => {
                        let ident = self.config.make_ident("PCWSTR", IdentKind::Type);
                        let ts = quote! { pub type #ident = *const u16; };
                        (ident, ts)
                    }
                    Type::BSTR => {
                        let ident = self.config.make_ident("BSTR", IdentKind::Type);
                        let ts = quote! { pub type #ident = *const u16; };
                        (ident, ts)
                    }
                    Type::GUID => {
                        let ident = self.config.make_ident("GUID", IdentKind::Record);
                        let impls = self.impls(&ident, Impls::COPY);

                        let ts = quote! {
                            #[repr(C)]
                            pub struct #ident {
                                pub data1: u32,
                                pub data2: u16,
                                pub data3: u16,
                                pub data4: [u8; 8],
                            }

                            #impls

                            impl #ident {
                                #[allow(dead_code)]
                                pub const fn from_u128(uuid: u128) -> Self {
                                    Self { data1: (uuid >> 96) as u32, data2: (uuid >> 80 & 0xffff) as u16, data3: (uuid >> 64 & 0xffff) as u16, data4: (uuid as u64).to_be_bytes() }
                                }
                            }
                        };
                        (ident, ts)
                    }
                    _ => {
                        continue;
                    }
                };

                os.insert_type(ty.clone(), ident, ts);
            } else if let Type::TypeDef((td, _)) = &ty {
                let def = *td;

                if reader.type_def_enclosing_type(def).is_some() {
                    continue;
                }

                let kind = reader.type_def_kind(def);
                let name = reader.type_def_name(def);

                let (ident, ts) = match kind {
                    TypeKind::Class => {
                        let ident = self.config.make_ident(name, IdentKind::Record);

                        let ts = if self.config.use_core {
                            //self.emit_class(&mut os, reader, def)
                            unreachable!()
                        } else {
                            quote! { pub type #ident = *mut ::core::ffi::c_void; }
                        };

                        (ident, ts)
                    }
                    TypeKind::Interface => {
                        let ident = self.config.make_ident(name, IdentKind::Record);

                        let ts = if self.config.use_core {
                            //self.emit_interface(&mut os, reader, def)
                            unreachable!()
                        } else {
                            quote! { pub type #ident = *mut ::core::ffi::c_void; }
                        };

                        (ident, ts)
                    }
                    TypeKind::Enum => {
                        os.insert_enum(ty.clone());
                        continue;
                    }
                    TypeKind::Struct => {
                        self.emit_record(&mut os, def, *impls)?;
                        continue;
                    }
                    TypeKind::Delegate => self.emit_function_pointer(def)?,
                };

                os.insert_type(ty.clone(), ident, ts);
            }
        }

        for constant in &self.items.constants {
            self.emit_constant(&mut os, *constant)?;
        }

        for func in &self.items.functions {
            self.emit_func(&mut os, *func);
        }

        let ts = os.finalize(self.config);

        if self.config.pretty_print {
            let file = syn::parse2(ts).context("unable to parse output as a valid Rust file")?;
            Ok(prettyplease::unparse(&file))
        } else {
            Ok(ts.to_string())
        }
    }
}

pub fn load_clang_layouts() -> &'static BTreeMap<String, ArchLayouts> {
    use std::sync::Once;

    const COMPRESSED_LAYOUTS: &[u8] = include_bytes!("../../md/layouts.json.zstd");
    static INIT: Once = Once::new();
    static mut LAYOUTS: Option<BTreeMap<String, ArchLayouts>> = None;

    unsafe {
        INIT.call_once(|| {
            let decompressed = zstd::decode_all(std::io::Cursor::new(COMPRESSED_LAYOUTS))
                .expect("failed to decompress layouts");
            LAYOUTS =
                Some(serde_json::from_slice(&decompressed).expect("failed to deserialize layouts"));
        });
        LAYOUTS.as_ref().expect("failed to deserialize layouts")
    }
}
