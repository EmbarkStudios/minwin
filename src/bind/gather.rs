use super::{InterfaceMap, ItemSet};
use std::collections::{BTreeMap, BTreeSet};
use windows_metadata::reader::{self, Reader, Type, TypeDef, TypeKind};

/// Takes a list of fully qualified type names and recursively gathers all of
/// the items needed to fully define them.
pub struct Gatherer<'names, 'r> {
    reader: &'r Reader<'r>,
    interfaces: &'names InterfaceMap,
    items: BTreeMap<&'names str, Item>,
}

impl<'names, 'r> Gatherer<'names, 'r> {
    pub fn new(
        reader: &'r Reader<'r>,
        items: impl Iterator<Item = (&'names str, Item)>,
        interfaces: &'names InterfaceMap,
    ) -> Self {
        Self {
            reader,
            interfaces,
            items: items
                .chain(
                    interfaces
                        .keys()
                        .map(|iname| (iname.as_str(), Default::default())),
                )
                .collect(),
        }
    }

    pub fn gather(self) -> ItemSet {
        let mut types = BTreeMap::new();
        let mut functions = BTreeSet::new();
        let mut constants = BTreeSet::new();

        let reader = self.reader;

        for (name, item) in &self.items {
            let type_name = reader::TypeName::parse(name);

            let mut found = false;

            if matches!(item.dis, Disambiguate::Any | Disambiguate::Record) {
                for def in reader.get(type_name) {
                    self.collect(&Type::TypeDef((def, vec![])), &mut types);
                    found = true;
                }
            }

            if found {
                continue;
            }

            if matches!(item.dis, Disambiguate::Any | Disambiguate::Function) {
                for method in reader
                    .namespace_functions(type_name.namespace)
                    .filter(|method| reader.method_def_name(*method) == type_name.name)
                {
                    functions.insert(method);
                    let signature = reader.method_def_signature(method, &[]);
                    signature
                        .return_type
                        .iter()
                        .for_each(|ty| self.collect(ty, &mut types));
                    signature
                        .params
                        .iter()
                        .for_each(|param| self.collect(&param.ty, &mut types));
                    found = true;
                }
            }

            if found || !matches!(item.dis, Disambiguate::Any | Disambiguate::Constant) {
                continue;
            }

            let field_type = |row: reader::Field| -> Type {
                let ft = reader.field_type(row, None);

                // Normally, this would be an HSTRING, but not for constants
                if ft == Type::String {
                    return if reader.field_is_ansi(row) {
                        Type::PCSTR
                    } else {
                        Type::PCWSTR
                    };
                }

                ft
            };

            if let Some(field) = reader
                .namespace_constants(type_name.namespace)
                .find(|field| reader.field_name(*field) == type_name.name)
            {
                constants.insert(field);
                let ty = field_type(field).to_const_type();
                self.collect(&ty, &mut types);
            }

            if let Some(field) = reader
                .namespace_types(type_name.namespace, &Default::default())
                .find_map(|def| {
                    if reader.type_def_kind(def) == TypeKind::Enum {
                        return reader
                            .type_def_fields(def)
                            .find(|field| reader.field_name(*field) == type_name.name);
                    }
                    None
                })
            {
                constants.insert(field);
                self.collect(&reader.field_type(field, None), &mut types);
            }
        }

        self.fixup_impls(&mut types);

        ItemSet {
            types,
            functions,
            constants,
        }
    }

    fn collect(&self, ty: &Type, collected: &mut BTreeMap<Type, Impls>) {
        // Get the underlying type, the pointers are irrelevant
        let ty = match ty {
            Type::MutPtr((ty, _))
            | Type::ConstPtr((ty, _))
            | Type::Win32Array((ty, _))
            | Type::WinrtArray(ty)
            | Type::WinrtArrayRef(ty)
            | Type::WinrtConstRef(ty) => &ty,
            _ => ty,
        };

        if collected.contains_key(ty) {
            return;
        }

        collected.insert(ty.clone(), Impls::default());

        let Type::TypeDef((def, generics)) = ty else { return; };
        let def = *def;
        let reader = self.reader;

        // Ensure that we get _all_ the types that match
        let type_name = reader.type_def_type_name(def);
        if !type_name.namespace.is_empty() {
            for row in reader.get(type_name) {
                if def != row {
                    self.collect(&Type::TypeDef((row, Vec::new())), collected);
                }
            }
        }

        for generic in generics {
            self.collect(&generic, collected);
        }

        for field in reader.type_def_fields(def) {
            let ty = reader.field_type(field, Some(def));
            self.collect(&ty, collected);
        }

        let kind = reader.type_def_kind(def);
        let ty_name = reader.type_def_name(def);

        let methods = if kind == TypeKind::Interface {
            let Some(im) = self.interfaces.get(ty_name) else { return; };
            Some(im)
        } else {
            None
        };

        for method in reader.type_def_methods(def) {
            // Skip delegate pseudo-constructors.
            let name = reader.method_def_name(method);
            if name == ".ctor" {
                continue;
            }

            if let Some(m) = &methods {
                if !m.contains(name) {
                    tracing::debug!("skipping {ty_name}::{name}");
                    continue;
                }
            }

            let sig = reader.method_def_signature(method, generics);
            if let Some(rt) = &sig.return_type {
                self.collect(rt, collected);
            }
            for param in &sig.params {
                self.collect(&param.ty, collected);
            }
        }

        for interface in reader.type_interfaces(&ty) {
            self.collect(&interface.ty, collected);
        }

        if kind == TypeKind::Struct
            && reader.type_def_fields(def).next().is_none()
            && reader.type_def_guid(def).is_some()
        {
            collected.insert(Type::GUID, Impls::empty());
        }

        self.collect_nested(def, collected);
    }

    fn collect_nested(&self, td: TypeDef, collected: &mut BTreeMap<Type, Impls>) {
        let reader = self.reader;
        for nested in reader.nested_types(td) {
            self.collect_nested(nested, collected);

            for field in reader.type_def_fields(nested) {
                let ty = reader.field_type(field, Some(nested));
                self.collect(&ty, collected);
            }
        }
    }

    /// Recursively adds implementations based upon the requested implementations
    /// for root items.
    ///
    /// We _could_ do this while collecting types, but it's kind of annoying and
    /// easier to just do it in a separate fixup pass
    fn fixup_impls(&self, collected: &mut BTreeMap<Type, Impls>) {
        for (name, impls) in self
            .items
            .iter()
            .filter_map(|(name, item)| (!item.ri.is_empty()).then_some((*name, item.ri)))
        {
            let type_name = reader::TypeName::parse(name);

            for def in self.reader.get(type_name) {
                self.fill_impls(Type::TypeDef((def, Vec::new())), impls, collected);
            }
        }
    }

    fn fill_impls(&self, ty: Type, impls: Impls, collected: &mut BTreeMap<Type, Impls>) {
        let ty = match ty {
            // We can skip fields that are behind a pointer
            Type::MutPtr(_) | Type::ConstPtr(_) => return,
            Type::Win32Array((ty, _))
            | Type::WinrtArray(ty)
            | Type::WinrtArrayRef(ty)
            | Type::WinrtConstRef(ty) => *ty,
            _ => ty,
        };

        let Type::TypeDef((def, _generics)) = &ty else { return; };
        let def = *def;
        let reader = self.reader;

        if let Some(imp) = collected.get_mut(&ty) {
            if imp.contains(impls) {
                // If all of the flags are already present we don't need to recurse the type
                return;
            }

            imp.insert(impls);
        }

        for field in reader.type_def_fields(def) {
            let ty = reader.field_type(field, Some(def));
            if let Type::TypeDef((fdef, _)) = &ty {
                if reader.type_def_namespace(*fdef).is_empty() {
                    continue;
                }
            }

            self.fill_impls(ty, impls, collected);
        }

        self.fill_impls_nested(def, impls, collected);
    }

    fn fill_impls_nested(&self, td: TypeDef, impls: Impls, collected: &mut BTreeMap<Type, Impls>) {
        let reader = self.reader;
        for nested in reader.nested_types(td) {
            self.fill_impls_nested(nested, impls, collected);
            self.fill_impls(Type::TypeDef((nested, Vec::new())), impls, collected);
        }
    }
}

/// Disambiguates items during gathering.
///
/// In [some](https://docs.rs/windows-sys/0.48.0/windows_sys/Win32/UI/Input/KeyboardAndMouse/struct.VK_F.html)
/// [cases](https://docs.rs/windows-sys/0.48.0/windows_sys/Win32/UI/Input/KeyboardAndMouse/constant.VK_F.html)
/// items with the exact same name can be located within the same namespace,
/// which can cause the incorrect item to be emitted and the desired item to be
/// skipped.
#[derive(Copy, Clone, Debug, Default, PartialEq, Eq)]
pub enum Disambiguate {
    /// The item is a constant or enum variant
    Constant,
    /// The item is a function
    Function,
    /// The item is a struct or union
    Record,
    #[default]
    Any,
}

bitflags::bitflags! {
    /// Decides additional implementations that are derived for the struct
    ///
    /// `windows-bindgen` unconditionally emits `Copy` and `Clone` for all records,
    /// but this increases compile times for:
    ///
    /// * No benefit because they are only ever moved
    /// * Worse, extremely large types are accidentally copied when normally the
    /// user would be given an error that it's been moved when doing trivial
    /// operations like accessing a single field etc.
    #[derive(Debug, Copy, Clone, Default)]
    pub struct Impls: u8 {
        const COPY = 1 << 0 | Impls::CLONE.bits();
        const CLONE = 1 << 1;
    }
}

#[derive(Default, Copy, Clone, Debug)]
pub struct Item {
    pub dis: Disambiguate,
    pub ri: Impls,
}

impl Item {
    pub(super) fn parse(bname: String) -> anyhow::Result<(Self, String)> {
        match (bname.find('-'), bname.find('+')) {
            (None, None) => Ok((Self::default(), bname)),
            (Some(_), Some(_)) => {
                anyhow::bail!("'{bname}' has both a disambiguation and record implementation, you can just specify the impl if you want a record");
            }
            (Some(i), None) => {
                let dis = match &bname[..i] {
                    "constant" => Disambiguate::Constant,
                    "function" => Disambiguate::Function,
                    "record" => Disambiguate::Record,
                    unknown => {
                        anyhow::bail!("unknown disambiguation prefix '{unknown}' for {bname}");
                    }
                };

                let stripped = bname[i + 1..].to_owned();

                Ok((
                    Self {
                        dis,
                        ..Default::default()
                    },
                    stripped,
                ))
            }
            (None, Some(i)) => {
                let mut ri = Impls::empty();

                for imp in bname[i + 1..].split('+') {
                    match imp {
                        "Copy" => ri.insert(Impls::COPY),
                        "Clone" => ri.insert(Impls::CLONE),
                        unknown => {
                            anyhow::bail!("unknown record impl '{unknown}' for {bname}");
                        }
                    };
                }

                let stripped = bname[..i].to_owned();

                Ok((
                    Self {
                        ri,
                        ..Default::default()
                    },
                    stripped,
                ))
            }
        }
    }
}
