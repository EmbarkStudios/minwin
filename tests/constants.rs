use minwin::bind;
use rstest::rstest;

/// Ensures we can generate string constants
#[rstest]
fn strings(
    #[values(
        bind::BindConfig::Minwin(Default::default()),
        // This is currently broken, see https://github.com/microsoft/windows-rs/issues/2499
        bind::BindConfig::Bindgen
    )]
    config: bind::BindConfig,
) {
    let name = config.to_string();
    let bo = bind::bind(
        [
            // utf-16
            "Windows.Win32.Devices.Enumeration.Pnp.ADDRESS_FAMILY_VALUE_NAME",
            // ansi
            "Windows.Win32.Media.Multimedia.JOY_CONFIGCHANGED_MSGSTRING",
        ],
        Default::default(),
        config,
    )
    .expect("failed to bind");

    insta::assert_snapshot!(format!("strings__{name}"), bo.bindings);
}
