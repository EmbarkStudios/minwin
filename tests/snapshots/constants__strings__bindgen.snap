---
source: tests/constants.rs
expression: bo.bindings
---
// Bindings generated by `windows-bindgen` 0.49.0

#![allow(
    non_snake_case,
    non_upper_case_globals,
    non_camel_case_types,
    dead_code,
    clippy::all
)]
pub const ADDRESS_FAMILY_VALUE_NAME: PCWSTR = w!("AddressFamily");
pub type HSTRING = *mut ::core::ffi::c_void;
pub const JOY_CONFIGCHANGED_MSGSTRING: PCSTR = s!("MSJSTICK_VJOYD_MSGSTR");

